python/cudf/cudf/core/scalar.py

```diff
@@ -70,13 +71,24 @@ def __init__(self, value, dtype=None):
                 self._host_dtype = value._host_dtype
             else:
                 self._device_value = value._device_value
-        elif isinstance(value, cudf._lib.scalar.DeviceScalar):
-            self._device_value = value
```

Comment:
I removed this just to identify any places we might be constructing from a `DeviceScalar` and it turns out there were not many at all.

Should we keep this change and disallow constructing from `DeviceScalar` altogether (in favour of the `from_device_scalar`  factory)?

Reply:
I would definitely be in favor of that if it's not too disruptive a change to make in this PR. Otherwise, we could make that change in a separate PR, merge that first, then merge this PR. Up to you! I definitely think that is the way to go, though. The `DeviceScalar` isn't part of the public API, so no need to handle it in a public object's constructor.

Reply:
Oh I guess you've already made all the necessary replacements. In that case, this PR looks great!

Reply:
It's probably out of scope of this PR but I wonder now if this might be something we're underutilizing internally. I see that for instance in general reductions we're taking the host value [right out of cython](https://github.com/rapidsai/cudf/blob/branch-22.08/python/cudf/cudf/_lib/reduce.pyx#L89) rather than wrapping it in a `cudf.Scalar`. I feel like the most consistent thing would be for scalars to exist as `cudf.Scalars` anywhere in our python internals any only become a python host value when they're being returned from a user facing API. I can imagine this could prevent a little copying here and there.

Reply:
@brandon-b-miller could you write up a separate issue with your suggestion here?